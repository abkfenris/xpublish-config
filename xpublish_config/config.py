import importlib

from typing import Dict, List, Optional, Union
from pathlib import Path

from goodconf import GoodConf, Field
from pydantic import create_model, Extra, BaseModel

from xpublish.plugins import find_default_plugins

class XpublishPluginSettings(GoodConf):
    disabled_plugins: List[str] = Field(
        default_factory=list, 
        description=(
            "A list of plugins to disable. "
            "The name is what it would normally be automatically loaded as."
        )
    )
    register_plugins: Dict[str, str] = Field(default_factory=list)

    class Config:
        env_prefix = "xpublish_"
        extra=Extra.ignore
        # env_nested_delimiter = '__'

class XpublishBaseSettings(XpublishPluginSettings, extra=Extra.forbid):
    pass


class RestSettings(BaseModel):
    # datasets: dict = Field(default_factory=dict)
    # routers: 
    cache_kws: dict = Field(default_factory=dict)
    app_kws: dict = Field(default_factory=dict)

class ServeSettings(BaseModel):
    host: str = Field(default="0.0.0.0")
    port: int = Field(default=9000)
    log_level: str = Field(default="info")
    debug: bool = Field(default=False)
    reload: bool = Field(default=False)


class XpublishServingSettings(XpublishBaseSettings):
    rest: RestSettings = Field(default_factory=RestSettings)
    serve: ServeSettings = Field(default_factory=ServeSettings)

    def rest_kwargs(self):
        kwargs = self.rest.dict()
        try:
            kwargs['plugins'] = self.plugins.__dict__
        except AttributeError:
            pass
        return kwargs


class XpublishConfigManager:
    def __init__(self, settings_class: Optional[XpublishBaseSettings] = None) -> None:
        if not settings_class:
            self.settings_class = XpublishServingSettings
        else:
            self.settings_class = settings_class

    def build_dynamic_settings(self, 
                               config: Optional[dict] = None, 
                               from_file: Optional[Union[str, Path]] = None):
        """Using the settings class, the included and excluded plugins
        return a settings class with plugins
        """
        plugin_settings = XpublishPluginSettings()
        plugin_settings.load(from_file)

        if config:
            plugin_settings = plugin_settings.copy(update=config)

        plugins = find_default_plugins(plugin_settings.disabled_plugins)

        for name, plugin_import_path in plugin_settings.register_plugins.items():
            try:
                lib_path, plugin_class_name = plugin_import_path.split(":")
            except ValueError as e:
                raise ValueError(f"Unable to build import path `{plugin_import_path}` when attempting to load `register_plugins.{name}. Is there a `:` missing?") from e
            
            plugin_lib = importlib.import_module(lib_path)

            try:
                plugin_class = getattr(plugin_lib, plugin_class_name)
            except AttributeError as e:
                raise AttributeError(f"{lib_path} doesn't appear to have have {plugin_class_name}. Maybe you meant one of these instead:\n{dir(plugin_lib)}") from e
        
            plugins[name] = plugin_class


        PluginModel = create_model(
            "PluginModel", 
            **{
                key: (value, Field(default_factory=value)) 
                for key, value in plugins.items()
                })

        class XpublishDynamicSettings(self.settings_class):
            plugins: PluginModel = Field(default_factory=PluginModel)

        return XpublishDynamicSettings

    def generate(self, type: str, config: Optional[dict] = None, from_file: Optional[Union[str, Path]] = None):
        """Create settings template files with initial config"""
        settings_class = self.build_dynamic_settings(config=config, from_file=from_file)

        if type.lower() in {"yaml", "yml"}:
            try:
                return settings_class.generate_yaml()
            except ImportError:
                raise ImportError("YAML generation requires ruamel to be installed.")
        
        if type.lower() == "json":
            return settings_class.generate_json()
        
        if type.lower() == "toml":
            try:
                return settings_class.generate_toml()
            except ImportError:
                raise ImportError("TOML generation requires tomlkit")
            
        if type.lower() in {"md", "markdown"}:
            return settings_class.generate_markdown()
        
        raise TypeError(f"{type} isn't a known type!")

    def parse(self, 
              config: Optional[dict] = None, 
              from_file: Optional[Union[str, Path]] = None
            ):
        """Return a fully configured settings class"""

        settings_class = self.build_dynamic_settings(config, from_file)
        settings = settings_class()

        settings.load(from_file)

        if config:
            settings = settings.copy(update=config)

        return settings
        

config = {
    # "disabled_plugins": ["info"]
    "register_plugins": {
        "test_local": "local_plugin:LocalPlugin"
    },
    "plugins": {
        "opendap": {
            "dataset_router_prefix": "/dap"
        }
    }
}


# print(XpublishConfigManager().generate("toml"))
# settings = XpublishConfigManager().parse(config=config)
# settings = XpublishConfigManager().parse(from_file="./test_config.yaml")

# print(settings)
# print(settings.rest_kwargs())