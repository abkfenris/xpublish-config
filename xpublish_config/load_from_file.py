"""Load a configuration from YAML, TOML, or JSON."""
from pathlib import Path
from typing import Union


def load_from_file(file_path: Union[Path, str]) -> dict:
    """Load a configuration from YAML, TOML, or JSON."""
    if isinstance(file_path, str):
        file_path = Path(file_path)

    if file_path.suffix in [".yml", ".yaml"]:
        import ruamel.yaml

        with file_path.open() as f:
            yaml = ruamel.yaml.YAML(typ="safe", pure=True).load(f)
            return yaml

    if file_path.suffix in [".toml"]:
        import tomlkit

        with file_path.open() as f:
            return tomlkit.load(f).unwrap()

    import json

    with file_path.open() as f:
        return json.load(f)
