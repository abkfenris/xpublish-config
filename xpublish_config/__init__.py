from .config import XpublishConfigManager

__all__ = ["XpublishConfigManager"]

try:
    from ._version import __version__
except ImportError:
    __version__ = "unknown"
