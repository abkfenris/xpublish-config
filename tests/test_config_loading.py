from xpublish_config import XpublishConfigManager

def test_load_settings_from_file(snapshot):
    settings = XpublishConfigManager().parse(from_file="./tests/test_config.yaml")

    assert settings.disabled_plugins == []
    assert settings.register_plugins == {
        "test_local": "a_local_plugin:TestLocalPlugin", 
        'gfs_datasets': 'xpublish_intake_provider:IntakeDatasetProviderPlugin'
    }

    assert settings.rest.cache_kws == {}
    assert settings.rest.app_kws == {}

    assert settings.serve.host == '0.0.0.0'
    assert settings.serve.port == 9000
    assert settings.serve.log_level == 'info'
    assert settings.serve.debug == False
    assert settings.serve.reload == False

    assert settings == snapshot


def test_disable_plugins_with_environment_variables(monkeypatch, snapshot):
    monkeypatch.setenv("XPUBLISH_DISABLED_PLUGINS", '["zarr"]')

    settings = XpublishConfigManager().parse(from_file="./tests/test_config.yaml")

    assert settings.disabled_plugins == ["zarr"]
    assert settings.register_plugins == {
        "test_local": "a_local_plugin:TestLocalPlugin", 
        'gfs_datasets': 'xpublish_intake_provider:IntakeDatasetProviderPlugin'
    }

    assert settings.rest.cache_kws == {}
    assert settings.rest.app_kws == {}

    assert settings.serve.host == '0.0.0.0'
    assert settings.serve.port == 9000
    assert settings.serve.log_level == 'info'
    assert settings.serve.debug == False
    assert settings.serve.reload == False

    assert settings == snapshot
