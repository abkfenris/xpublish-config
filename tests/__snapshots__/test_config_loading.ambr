# serializer version: 1
# name: test_disable_plugins_with_environment_variables
  dict({
    'disabled_plugins': list([
      'zarr',
    ]),
    'plugins': dict({
      'cf_edr': dict({
        'app_router_prefix': '/edr',
        'app_router_tags': list([
          'edr',
        ]),
        'dataset_router_prefix': '/edr',
        'dataset_router_tags': list([
          'edr',
        ]),
        'name': 'cf_edr',
      }),
      'dataset_info': dict({
        'dataset_router_prefix': '',
        'dataset_router_tags': list([
          'dataset_info',
        ]),
        'name': 'dataset_info',
      }),
      'gfs_datasets': dict({
        'name': 'intake-dataset-provider',
        'uri': 'https://raw.githubusercontent.com/axiom-data-science/mc-goods/main/mc_goods/gfs-1-4deg.yaml',
      }),
      'module_version': dict({
        'app_router_prefix': '',
        'app_router_tags': list([
          'module_version',
        ]),
        'name': 'module_version',
      }),
      'opendap': dict({
        'dataset_router_prefix': '/opendap',
        'dataset_router_tags': list([
          'opendap',
        ]),
        'name': 'opendap',
      }),
      'plugin_info': dict({
        'app_router_prefix': '',
        'app_router_tags': list([
          'plugin_info',
        ]),
        'name': 'plugin_info',
      }),
      'test_local': dict({
        'name': 'local_test',
      }),
    }),
    'register_plugins': dict({
      'gfs_datasets': 'xpublish_intake_provider:IntakeDatasetProviderPlugin',
      'test_local': 'a_local_plugin:TestLocalPlugin',
    }),
    'rest': dict({
      'app_kws': dict({
      }),
      'cache_kws': dict({
      }),
    }),
    'serve': dict({
      'debug': False,
      'host': '0.0.0.0',
      'log_level': 'info',
      'port': 9000,
      'reload': False,
    }),
  })
# ---
# name: test_load_settings_from_file
  dict({
    'disabled_plugins': list([
    ]),
    'plugins': dict({
      'cf_edr': dict({
        'app_router_prefix': '/edr',
        'app_router_tags': list([
          'edr',
        ]),
        'dataset_router_prefix': '/edr',
        'dataset_router_tags': list([
          'edr',
        ]),
        'name': 'cf_edr',
      }),
      'dataset_info': dict({
        'dataset_router_prefix': '',
        'dataset_router_tags': list([
          'dataset_info',
        ]),
        'name': 'dataset_info',
      }),
      'gfs_datasets': dict({
        'name': 'intake-dataset-provider',
        'uri': 'https://raw.githubusercontent.com/axiom-data-science/mc-goods/main/mc_goods/gfs-1-4deg.yaml',
      }),
      'module_version': dict({
        'app_router_prefix': '',
        'app_router_tags': list([
          'module_version',
        ]),
        'name': 'module_version',
      }),
      'opendap': dict({
        'dataset_router_prefix': '/opendap',
        'dataset_router_tags': list([
          'opendap',
        ]),
        'name': 'opendap',
      }),
      'plugin_info': dict({
        'app_router_prefix': '',
        'app_router_tags': list([
          'plugin_info',
        ]),
        'name': 'plugin_info',
      }),
      'test_local': dict({
        'name': 'local_test',
      }),
      'zarr': dict({
        'dataset_router_prefix': '/zarr',
        'dataset_router_tags': list([
          'zarr',
        ]),
        'name': 'zarr',
      }),
    }),
    'register_plugins': dict({
      'gfs_datasets': 'xpublish_intake_provider:IntakeDatasetProviderPlugin',
      'test_local': 'a_local_plugin:TestLocalPlugin',
    }),
    'rest': dict({
      'app_kws': dict({
      }),
      'cache_kws': dict({
      }),
    }),
    'serve': dict({
      'debug': False,
      'host': '0.0.0.0',
      'log_level': 'info',
      'port': 9000,
      'reload': False,
    }),
  })
# ---
